Openshift checkouts do not include git submodule trees

Related to #320, Openshift 3.9 is not copying the submodules, causing Mantle builds to fail.

Example:
````
+ test -f mantle/README.md
+ '[' -n '' ']'
+ '[' -n '' ']'
+ make
cd mantle && ./build ore kola kolet
fatal: No names found, cannot describe anything.
make: *** [Makefile:12: mantle] Error 128
Removing intermediate container 3bfa9eaa6497
error: build error: The command '/bin/sh -c ./build.sh make_and_makeinstall' returned a non-zero code: 2
```

Further, the test for `mantle/README.md` is insufficient since the build checks the status for the Git tree.

--- a/build.sh
+++ b/build.sh
@@ -84,13 +84,6 @@ install_rpms() {
 }
 
 _prep_make_and_make_install() {
-    # Work around https://github.com/coreos/coreos-assembler/issues/27
-    if ! test -d .git; then
-        (git config --global user.email dummy@example.com
-         git init && git add . && git commit -a -m 'dummy commit'
-         git tag -m tag dummy-tag) >/dev/null
-    fi
-
     # TODO: install these as e.g.
     # /usr/bin/ostree-releng-script-rsync-repos
     mkdir -p /usr/app/

--- a/build.sh
+++ b/build.sh
@@ -96,7 +96,7 @@ _prep_make_and_make_install() {
     mkdir -p /usr/app/
     rsync -rlv "${srcdir}"/ostree-releng-scripts/ /usr/app/ostree-releng-scripts/
 
-    if ! test -f mantle/README.md; then
+    if [ "$(git submodule status mantle | head -c1)" == "-" ]; then
         echo -e "\033[1merror: submodules not initialized. Run: git submodule update --init\033[0m" 1>&2
         exit 1
     fi

--- a/src/cmdlib.sh
+++ b/src/cmdlib.sh
@@ -299,11 +299,21 @@ openshift_git_hack() {
 
     # Some versions of Openshift do not include the GIT repo in the checkout.
     # This may be needed for Openshift versions 3.9 and earlier.
-    # See https://github.com/coreos/coreos-assembler/issues/320.
+    # See https://github.com/coreos/coreos-assembler/issues/320 and
+    #      https://github.com/coreos/coreos-assembler/pull/341
+    #
+    # Due to limitations in RHEL 7, using "GIT_WORK_TREE" options does not work with
+    # submodules. As a result, a new checkout is done, which is rsynced into the
+    # source; rsync is less than ideal, but it allows a work-around for the combination
+    # of RHEL 7 and problematic Openshift versions.
 
-    # $OPENSHIFT_GIT_HACK, $GIT_REF and $GIT_URL are environment variables,
-    # provided by Openshift or the container run time. See
-    # https://github.com/coreos/coreos-assembler/pull/324 for an example.
+    # This hack SHOULD ONLY BE HIT WHEN RUNNING IN A CI SITUATION, under
+    # the following assumptions:
+    #  - The source being built originated from a clean git checkout.
+    #  - The `.git` repository has been stripped out by CI.
+    #  - The envVar parameters of $OPENSHIFT_GIT_HACK, $GIT_REF, and $GIT_URL are
+    #    defined, signalling that this code is intended to run. See
+    #    https://github.com/coreos/coreos-assembler/pull/324 for an example.
 
     local gitd=${1}; shift;
 
@@ -318,8 +328,21 @@ openshift_git_hack() {
     info "NOTICE: using workaround for Openshift missing .git"
 
     if [ "${GIT_URL:-x}" == "x" ] ||  [ "${GIT_REF:-x}" ]; then
-        info "Checking out bare git of ${GIT_URL} to ${gitd}/.git, ref ${GIT_REF}"
-        git clone --depth 1 --bare -b "${GIT_REF}" "${GIT_URL}" "${gitd}/.git"
+        tmpgit="$(mktemp -d)"
+        trap 'rm -rf ${tmpgit}; unset tmpgit;' EXIT
+
+        info "Re-creating git tree in ${gitd} from ${GIT_URL}, ref ${GIT_REF}"
+        git clone --depth 1 -b "${GIT_REF}" "${GIT_URL}" "${tmpgit}/clone"
+
+        if test -f "${gitd}/.gitmodules"; then
+            pushd "${tmpgit}/clone"
+            info "Ensuring git modules exists too"
+            git submodule update --init --force
+            popd
+        fi
+
+        # For more context see the discussion in https://github.com/coreos/coreos-assembler/pull/341
+        rsync -av --ignore-existing "${tmpgit}/clone/" "${gitd}"/
     fi
 }
 

